### ISSUE TITLE
[go_router] Redirection with Riverpod example

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#112915]
This is a request to add an example to the [`go_router`](https://pub.dev/packages/go_router) package that shows how to set up redirection with Riverpod.

## Use case

`go_router` is currently not very well documented and so it took me (as an individual who is familiar with how the different parts of the framework work) quite a while to figure out a solid way to use `go_router` with `flutter_riverpod` properly to handle a **sign-in flow**.

## Proposal

Because of that, I feel like the [existing example catalog](https://github.com/flutter/packages/blob/4bd1dc598aaab5f8179a5429986c36852a54dd63/packages/go_router/example/README.md) should be updated with another sample to complement "Redirection" and "Asynchronous Redirection".  
This is mostly because it works fairly differently under the hood.

---

Is this something you would be willing to integrate into the package? I would also be willing to contribute the example.  
I think this would be very beneficial to also ensure that updates to the package do not break this setup.

### Setup

This method makes use of the fact that reconstructing `GoRouter` will call `redirect` again.  
That means that by providing a `redirect` function that depends on providers, we can both handle calls to `redirect` caused by internal calls (`Navigator.pop`, `context.go`, etc.) as well as the providers updating.

It looks something like this:

```dart
// This is super important - otherwise, we would throw away the whole widget tree when the provider is updated.
final navigatorKey = GlobalKey<NavigatorState>();

final routerProvider = Provider((ref) {
  return GoRouter(
    navigatorKey: navigatorKey,
    routes: [...],
    redirect: (context, state) {
      // if the user is not logged in, they need to login
      final bool loggedIn = ref.watch(loggedInProvider);
      final bool loggingIn = state.subloc == '/login';
      if (!loggedIn) {
        return loggingIn ? null : '/login';
      }

      // if the user is logged in but still on the login page, send them to
      // the home page
      if (loggingIn) {
        return '/';
      }

      // no need to redirect at all
      return null;
    },
  );
});
```

This setup is so much cleaner and easier to grasp than the [existing Redirection example using `provider`](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/redirection.dart).

Also note that I was not able to set this up using `refreshListenable`, which is not documented, confusing, and also does not work correctly from what I can tell (it can cause invalid states). Furthermore, the `refreshListenable` setup seems way more hacky than the `riverpod` setup above to begin with.

In conclusion, I think that this method is pretty powerful and I would like to see it included in the package publicly somehow in order to ensure that others can use it and it can be maintained (although it should always work if the package is implemented correctly and checks for equality properly without throwing away the widget tree).

**Comment by TahaTesser** [comment]
@creativecreatorormaybenot 
Do you plan to add this example yourself? If not, I can take this.
(Since Riverpod is supposed to be a better version of Provider, I think adding Riverpod examples for both "Redirection" and "Asynchronous Redirection" make sense to me)

**Comment by lucavenir** [comment]
Hi @creativecreatorormaybenot 

On [Riverpod's documentation pages](https://riverpod.dev/docs/getting_started), under the 3rd party examples folder, there's [a repo I've made a while ago](https://github.com/lucavenir/go_router_riverpod) you can explore to check out possible solutions to integrate GoRouter and Riverpod together.

It's going to be updated with GoRouter v5 + Riverpod v2 as soon as [this issue](https://github.com/flutter/flutter/issues/112798) will be fixed.

**Comment by creativecreatorormaybenot** [comment]
> Do you plan to add this example yourself? If not, I can take this.

@TahaTesser That would be super awesome! I certainly think that having Riverpod examples makes sense 👍 

I would be glad to review 👌

**Comment by creativecreatorormaybenot** [comment]
Hi @lucavenir - I have explored that solution previously as well. But to me that was even more of a hack and did not work perfectly.

> In your example snippet above, routerProvider will rebuild everytime there's an authentication change; this will most likely cause a whole app rebuild (onto the root of our apps), since I'd expect the main App to watch our routerProvider. This is most likely undesired.

This is not right. If you look at the snippet, you will see that I consciously pass a `navigatorKey` to `GoRouter`. This means that the tree will be **kept** upon "rebuilding" the `routerProvider`. This makes use of Flutter's ability to keep the tree by comparing its elements. Because the setup is the exact same before and after, the same widgets will be inserted and if you write your widgets correctly, all state will be kept and nothing will be discarded.

> The above (and more caveats) are what motivated [this repo](https://github.com/lucavenir/go_router_riverpod) to not only write down a simple example, but to properly analyze different possible approaches.

I very highly appreciate this effort 🙏 Thanks @lucavenir. However, I am sorry to tell you that I do not like the solutions in this repo. Especially piping all provider changes into listenables seems awful to maintain.

Feel free to try out my proposed solution and report if you can find any problems with it. This is also why I would really love to have an `example/` for this setup in `go_router` because we could properly test if there are any problems with state being discarded or undesired rebuilds.

### About rebuilds

Note that you are of course right that the `App` in this setup:

```dart
class App extends ConsumerWidget {
  const App({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final router = ref.watch(routerProvider);
    return MaterialApp.router(
      routerConfig: router,
    );
  }
}
```

Will always rebuild when one of the dependencies of the `routerProvider` changes. **However**, since we pass the `navigatorKey`, we really do not care about this. This rebuild basically has minimal cost (and the framework could also trigger it at any time) and all of our actual tree, i.e. our pages returned from the router, that we may not want to rebuild unnecessarily will **not** rebuild. That is because the framework checks if the pages have updated and they will not update unless there is an actual change in navigation.

I do, however, agree that `GoRouter` could be further optimized for being reconstructed. One crazy idea I have is removing `refreshListenable` altogether and recommending to rebuild `GoRouter` using a `ValueListenableBuilder` instead (or use `flutter_riverpod`). To me this seems so much cleaner and would also remove a bunch of complexity.

**Comment by creativecreatorormaybenot** [comment]
One crazy idea I have is removing `refreshListenable` altogether and recommending to rebuild `GoRouter` using say a `ValueListenableBuilder` instead (or using `flutter_riverpod` as outlined). To me this seems so much cleaner and would also remove a bunch of complexity.

If we further optimize `GoRouter` for being reconstructed, this would be the perfect solution for redirection.

**Comment by creativecreatorormaybenot** [comment]
One crazy idea I have is removing `refreshListenable` altogether and recommending to rebuild `GoRouter` using say a `ValueListenableBuilder` instead (or using `flutter_riverpod` as outlined). To me this seems so much cleaner and would also remove a bunch of complexity.

If we further optimize `GoRouter` for being reconstructed, this would be the perfect solution for redirection 🚀

**Comment by creativecreatorormaybenot** [comment]
One crazy idea I have is *removing `refreshListenable` altogether* and recommending to rebuild `GoRouter` instead. Say using `ValueListenableBuilder` or using `flutter_riverpod` as outlined.  To me this seems so much cleaner and would also **reduce** complexity.

If we further optimize `GoRouter` for being reconstructed, this would be the perfect solution for redirection 🚀

**Comment by lucavenir** [comment]
> Hi @lucavenir, this is not right. If you look at the snippet, you will see that I consciously pass a `navigatorKey` to `GoRouter`. This means that the tree will be **kept** upon "rebuilding" the `routerProvider`. This makes use of Flutter's ability to keep the tree by comparing its elements. Because the setup is the exact same before and after, the same widgets will be inserted and if you write your widgets correctly, all state will be kept and nothing will be discarded.
> Feel free to try out my proposed solution and report if you can find any problems with it.

I've been sloppy in reading your code, sorry if I made you lose time! Thanks for the detailed explanation.
This makes me rethink a lot of the integrations with Riverpod: we could just use `Provider`s like we always do: we watch elements in there declaratively, without implementing external `Listenable`s. I have to test this out and evaluate its scalability and if it's feasible (the Flutter tree won't rebuild - but our Provider containing the Router will).

> Especially piping all provider changes into listenables seems awful to maintain.

I'm obsessed with best practices and I can't agree more! I'm planning to erase out all the "not-so-good" practices in some examples. You see, those examples were community - driven, they came out from a [discussion](https://github.com/rrousselGit/riverpod/discussions/1357) in Riverpod's repo. Most of the examples feel like "legay code" ad this point. I've just released an update about this. I'm trying to maintain that repo as a "community wiki", so there's a good chance I'll touch the examples a lot in the next days.

> This is also why I would really love to have an `example/` for this setup in `go_router` 

I absolutely agree with this. See the above sentences.

**Comment by lucavenir** [comment]
@creativecreatorormaybenot Do you mind take a look at [this branch](https://github.com/lucavenir/go_router_riverpod/tree/alternative_approach)? I'm trying your suggestion but I still seeing the "root widget" rebuilding, there.

**Comment by creativecreatorormaybenot** [comment]
@lucavenir I am not sure what the "root widget" in your setup means, but here you go: https://dartpad.dev/0a77fd6105aeee44858b794e17e83ffc  
This is a full setup using my `riverpod` + `go_router` approach.

Here is an in-depth showcase 💯  Here we have..


https://user-images.githubusercontent.com/19204050/194175676-3c4b94b6-fad8-4999-a94a-7a54701211fb.mp4



Note that the `MaterialApp` itself also does not rebuild internally because it does not change. So it is mainly the `App.build` call, which is very cheap.

**Comment by lucavenir** [comment]
Thank you @creativecreatorormaybenot for the in-depth explanation!! This is way clearer. I guess I'm still confused by Flutter's `key`s.

**Comment by lucavenir** [comment]
Hi @creativecreatorormaybenot !

I've updated the examples in [this repo](https://github.com/lucavenir/go_router_riverpod), thanks to you! I've cited you as a contributor, but I'm unsure on how to give you more credit than that 😛

Your feedback would be highly appreciated - and I'm guessing that the `complete_example` folder in there could be used to solve this issue 👀

**Comment by creativecreatorormaybenot** [comment]
I really appreciate the sample @lucavenir. The crediting is fine. I think a link to the issue would help for context 🙂 

Note that I discovered a problem. In the case that you do not cover all of your routes in your `redirect` handler (e.g. if you have nested routes and only redirect to and from the top route), rebuilding the router will mess up your current state as the initial location is always `/`.

The workaround for this is adding:

```dart
...
// We need to have access to the previous location of the router. Otherwise, we would start from '/' on rebuild.
GoRouter? _previousRouter;

final routerProvider = Provider((ref) {
  return GoRouter(
    initialLocation: _previousRouter?.location,
    ...,
  );
});
```

I am not the biggest fan of this since it is even less clean (with the router being outside of any providers). However, it does work and is solid / non-fragile in theory.

Another implication is that modals will never work with this approach, i.e. routes that are pushed via `GoRouter.of(context).push` since they are not encapsulated in the current location. What does work is modals inside of a page, e.g. [`modal_bottom_sheet`](https://pub.dev/packages/modal_bottom_sheet). So this is not an issue.  
For routes that are pushed on top of the whole stack / not part of the routes defined in `routes`, I have come up with my own solution that integrates way better with `riverpod` anyway.

That is to say, I still greatly prefer this method over the previous methods. Although, we should be aware of the limitations.

---

I really like your samples - good job! They are not complex enough to have encountered the edge cases I mentioned, so it might be worth playing with it a bit more to discover these as well 👍

**Comment by Neelansh-ns** [comment]
> I am not sure what the "root widget" in your setup means @lucavenir, but here you go: **https://dartpad.dev/0a77fd6105aeee44858b794e17e83ffc** This is a full setup using my `riverpod` + `go_router` approach.
> 
> Here is an in-depth showcase 💯 Here we have (sound on 🔈)..
> 
>  Screen.Recording.2022-10-05.at.22.19.27.mp4 
> Note that the `MaterialApp` itself also does not rebuild internally because it does not change. So it is mainly the `App.build` call, which is very cheap.

This solution is not working as shown in the video on the latest stable channel `3.7.3`. @creativecreatorormaybenot

**Comment by creativecreatorormaybenot** [comment]
It may be that you skipped one of the steps I described @Neelansh-ns. I just tested the Gist at https://dartpad.dev/0a77fd6105aeee44858b794e17e83ffc and it works for me.

**Comment by lucavenir** [comment]
Hi there,

I want to share some thoughts.

1. I feel this example is very much needed. I'm still pretty much proud (😂) of the examples from my repo, but maybe having an "official" go-to solution is best;
2. After some trials, and after `riverpod v2`, together with `riverpod_generator`, I like [this solution](https://github.com/lucavenir/go_router_riverpod/blob/master/code_gen/lib/router/router_notifier.dart) better. In a nutshell: there's no passing of a `ref` object, `router` isn't rebuilding on authentication changes and our `notifier` actually implements `Listenable`, which is what's required by current's `go_router`'s API.
3. All of this conversation might be overridden by [this comment](https://github.com/flutter/flutter/issues/116651#issuecomment-1343462316) from `go_router` maintainers. I have a gut feeling that there's a good chance this issue is being "delayed" for a good reason: `go_router` is still on its way of a "best" design;
4. Therefore, taking all of this into account (i.e. `go_router` APIs have changed in a way that is making the maintainers think 'refreshListenable` might be deprecated) and adding [this other comment](https://github.com/flutter/flutter/issues/116651#issuecomment-1361716448) to the equation, maybe [this solution](https://github.com/flutter/flutter/issues/116651#issuecomment-1428028054) might be a good reference. Note: I don't like this very much, I don't find it declarative and I don't like the fact the responsibility of triggering redirects is moved away from `router`.

**Comment by rirjkl19** [comment]
@creativecreatorormaybenot Would like to ask on this. I actually liked your provided solution  as everything make sense in first look. However,

> I have come up with my own solution that integrates way better with `riverpod` anyway.

What was your solution on handling modals on top of the route stack?

And as for this. How do you update the `_previousRouter` then?

``` dart
...
// We need to have access to the previous location of the router. Otherwise, we would start from '/' on rebuild.
GoRouter? _previousRouter;

final routerProvider = Provider((ref) {
  return GoRouter(
    initialLocation: _previousRouter?.location,
    ...,
  );
});
```

**Comment by julgog** [comment]
[creativecreatorormaybenot](https://github.com/creativecreatorormaybenot) 
@creativecreatorormaybenot hello , sorry but your example only works for the web platform, testing on Android the widgets are rebuilt despite having the navigatorkey.
I think that even though it is rebuilt, the changes in authentication occur at startup and when closing, so in the worst case it would only be rebuilt twice while using the application.
Anyway I would like a cleaner solution if anyone has an idea how to use go_router with firebase

**Comment by mvarendorff** [comment]
A note on `_previousRouter?.location`; [since GoRouter 9.0.0](https://docs.google.com/document/d/16plvWc9ablQsUs7w6bWDpTZ7PwMP4YUhV-qMQ3iljE0/edit), it should be `_previousRouter?.routerDelegate.currentConfiguration.fullPath`.

**Comment by CesareIsHere** [comment]
Hi there,

I solved this problem by creating a router provider that who takes care about update the router every time the auth state changes.

I did some tests but i'm not sure if is conceptually right or not, so I would appreciate it if you could give me some feedback.

RouterNotifier:
```
@riverpod
class RouterNotifier extends _$RouterNotifier {
  @override
  FutureOr<GoRouter> build(WidgetRef reff) {
    ref.watch(authNotifierProvider);
    return routerGenerator(
      ref: reff,
      rootNavigatorKey: rootNavigatorKey,
      shellNavigatorKey: shellNavigatorKey,
    );
  }
}
```

In the MaterialApp i have just added the watcher to the gorouter notifier value
```
MaterialApp.router(
        title: 'App',
        debugShowCheckedModeBanner: false,
        routerConfig: ref.watch(routerNotifierProvider(ref)).value,
      )
```

Than this is the function that create the GoRouter object
```
GoRouter routerGenerator({
  required WidgetRef ref,
  GlobalKey<NavigatorState>? rootNavigatorKey,
  GlobalKey<NavigatorState>? shellNavigatorKey,
}) {
  return GoRouter(
    navigatorKey: rootNavigatorKey,
    initialLocation: PostsPage.route,
    redirect: (context, state) {
      if (!ref.read(authNotifierProvider.notifier).isSignedIn()) {
        return LoginPage.route;
      }
      return null;
    },
    routes: ...
```

Thank you for your feedback!

**Comment by BenjiFarquhar** [comment]
I was getting weird bugs when my `GoRouter` object reference changed due to using `ref.watch` inside the router provider. I think it's better to use `ref.read` inside the redirect method, so your `GoRouter` object reference is always the same, and listen to the auth status provider in the root widget and manually redirect based on the new auth status.
